wagtail/images/blocks.py

```diff
@@ -51,3 +57,148 @@ def htmldiff(self):
                 "image_b": self.val_b,
             },
         )
+
+
+class ImageBlock(StructBlock):
+    image = ImageChooserBlock(required=True)
+    decorative = BooleanBlock(
+        default=False, required=False, label=_("Image is decorative")
+    )
+    alt_text = CharBlock(required=False)
+
+    def get_searchable_content(self, value):
+        return []
+
+    def _struct_value_to_image(self, struct_value):
+        image = struct_value.get("image")
+        decorative = struct_value.get("decorative")
+        if image:
+            # If the image is decorative, set alt_text to an empty string
+            image.contextual_alt_text = (
+                "" if decorative else struct_value.get("alt_text", "")
+            )
+            image.decorative = decorative
+        return image
+
+    def to_python(self, value):
+        # For backward compatibility with ImageChooserBlock
+        if isinstance(value, int):
+            from wagtail.images import get_image_model
+
+            Image = get_image_model()
+            image = Image.objects.get(
+                id=value
+            )  # To retrieve specific image instance from db
+            struct_value = {"image": image, "decorative": False, "alt_text": ""}
+        else:
+            struct_value = super().to_python(value)
+        return self._struct_value_to_image(struct_value)
+
+    def bulk_to_python(self, values):
+        # For backward compatibility with ImageChooserBlock
+        struct_values = []
+        for value in values:
+            if isinstance(value, int):
+                from wagtail.images import get_image_model
+
+                Image = get_image_model()
+                image = Image.objects.get(
+                    id=value
+                )  # To retrieve specific image instance from db
```

Comment:
So here as well it seems to me like we need to re-implement. The point of the `bulk_to_python` methods as I understand them is reducing the number of queries by combining DB lookups. Unless we’ve agreed the optimization wasn’t worth chasing?

If it’s not worth chasing – then I’d expect this "bulk" method needs re-implementing to reuse the `to_python` method. Not worth duplicating the code as much as we are currently.

If it is worth chasing, then, seems like we need to use `ImageChooserBlock` / `FieldBlock`’s `bulk_to_python`. Convert our mixed `values` with direct image references and ints to one list of direct image references. Bulk fetch that. Then return with our other StructBlock fields’ data, and appropriate defaults for any image references.